dbms/src/Storages/Page/V3/WALStore.cpp

```diff
@@ -135,32 +142,42 @@ std::tuple<std::unique_ptr<LogWriter>, LogFilename> WALStore::createLogWriter(
         new_log_lvl.first,
         /*recycle*/ true,
         /*manual_flush*/ manual_flush);
-    return {
-        std::move(log_writer),
-        log_filename};
+    return {std::move(log_writer), log_filename};
 }
 
-WALStore::FilesSnapshot WALStore::getFilesSnapshot() const
+WALStore::FilesSnapshot WALStore::tryGetFilesSnapshot(size_t max_persisted_log_files, bool force)
 {
-    const auto [ok, current_writing_log_num] = [this]() -> std::tuple<bool, Format::LogNumberType> {
-        std::lock_guard lock(log_file_mutex);
-        if (!log_file)
+    // First we simply check whether the number of files is enough for compaction
+    LogFilenameSet persisted_log_files = WALStoreReader::listAllFiles(delegator, logger);
+    if (!force && persisted_log_files.size() <= max_persisted_log_files)
+    {
+        return WALStore::FilesSnapshot{};
+    }
+
+    SYNC_FOR("before_WALStore::tryGetFilesSnapshot_getLastPersistedLogNum");
+
+    Format::LogNumberType current_writing_log_num = 0;
+    {
+        std::lock_guard lock(log_file_mutex); // block other writes
+        if (log_file == nullptr && !force)
         {
-            return {false, 0};
+            // `log_file` is empty means there is no new writes
+            // after WALStore created. Just return an invalid snapshot
+            // if `force == false`.
+            return WALStore::FilesSnapshot{};
         }
-        return {true, log_file->logNumber()};
-    }();
-    // Return empty set if `log_file` is not ready
-    if (!ok)
-    {
-        return WALStore::FilesSnapshot{
-            .current_writing_log_num = 0,
-            .persisted_log_files = {},
-        };
+        // Reset the `log_file` so that next edit will be written to a
+        // new file and update the `last_log_num`
+        current_writing_log_num = last_log_num;
+        log_file.reset();
     }
 
-    // Only those files are totally persisted
-    LogFilenameSet persisted_log_files = WALStoreReader::listAllFiles(delegator, logger);
+    // Scan the log file list after we confirm the `last_persisted_log_num`,
+    // cause there could be new log files created between the last `listAllFiles`
+    // and acquiring `log_file_mutex`.
+    // We need to compact all log files before `last_persisted_log_num` so
+    // that it contains all entries changes we need.
+    persisted_log_files = WALStoreReader::listAllFiles(delegator, logger);
```

Comment:
This is not really necessary because GC is single-threaded. So the heavy gc won't run concurrently when dump snapshot. So `persisted_log_files` get at the start of this function is enough. New log files generated by foreground write is out of our concern.

Reply:
Removed